cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/lineage/field/DefaultFieldLineageReader.java

```diff
@@ -89,22 +93,42 @@ public DefaultFieldLineageReader(DatasetFramework datasetFramework, TransactionS
   }
 
   @Override
-  public Set<ProgramRunOperations> getIncomingOperations(EndPoint endPoint, long start, long end) {
-    return Transactionals.execute(transactional, context -> {
-      FieldLineageDataset fieldLineageDataset = FieldLineageDataset.getFieldLineageDataset(context, datasetFramework,
-                                                                                           fieldLineageDatasetId);
-
-      return fieldLineageDataset.getIncomingOperations(endPoint, start, end);
-    });
+  public Set<ProgramRunOperations> getIncomingOperations(EndPoint endPoint, String field, long start, long end) {
+    return computeFieldOperations(true, endPoint, field, start, end);
   }
 
   @Override
-  public Set<ProgramRunOperations> getOutgoingOperations(EndPoint endPoint, long start, long end) {
-    return Transactionals.execute(transactional, context -> {
+  public Set<ProgramRunOperations> getOutgoingOperations(EndPoint endPoint, String field, long start, long end) {
+    return computeFieldOperations(false, endPoint, field, start, end);
+  }
+
+  private Set<ProgramRunOperations> computeFieldOperations(boolean incoming, EndPoint endPoint, String field,
+                                                           long start, long end) {
+    Set<ProgramRunOperations> endPointOperations = Transactionals.execute(transactional, context -> {
       FieldLineageDataset fieldLineageDataset = FieldLineageDataset.getFieldLineageDataset(context, datasetFramework,
-                                                                                           fieldLineageDatasetId);
+              fieldLineageDatasetId);
 
-      return fieldLineageDataset.getOutgoingOperations(endPoint, start, end);
+      return incoming ? fieldLineageDataset.getIncomingOperations(endPoint, start, end)
+              : fieldLineageDataset.getOutgoingOperations(endPoint, start, end);
     });
+
+    Set<ProgramRunOperations> endPointFieldOperations = new HashSet<>();
+    for (ProgramRunOperations programRunOperation : endPointOperations) {
+      try {
+        FieldLineageInfo info = new FieldLineageInfo(programRunOperation.getOperations());
+        Set<Operation> fieldOperations = incoming ?
+                info.getIncomingOperationsForField(new EndPointField(endPoint, field))
```

Comment:
We store all operations as column in dataset. I wanted to minimize the work we do inside transaction (to avoid timeouts). So dataset simply returns all operations without doing any processing. FieldLineageReader then does the post processing on them to return the subset.

Reply:
yes that makes sense.